### ISSUE TITLE
Rule suggestion: prefer object spread to Object.assign

### ISSUE THREAD

**Issue by danvk** [palantir/tslint#2618]
Now that TS supports the object-spread operator, the idea would be to suggest replacing code like this:

```ts
Object.assign({}, foo, bar);
```
with

```ts
{...foo, ...bar}
```

See https://github.com/airbnb/javascript#objects--rest-spread

**Comment by ajafff** [comment]
@danvk You can already do that with the `ban` rule and add an appropriate message. This doesn't have automatic fixing though.

On second thought using `ban` would also disallow cases where you really want to assign properties of one object to another without creating a new one.

**Comment by danvk** [comment]
Yeah, this has slightly more complex semantics than "ban" would allow. If
it's specifically mentioned in the AirBnb guide, I think it warrants a rule.

**Comment by etsuo** [comment]
For what it's worth, an Object Spread does not appear to be the exact equivalent of Object.assign:
```
value = Object.assign(new Model(), value)
```
In this case, `value instanceOf Model` is true.
```
value = {...new Model(), ...value};
```
Here, value is an instance of Object.

This auto-fix just broke a bunch of my unit-tests... no biggie... just turned it off, but I figured I'd mention it in case someone thought the two were exactly equivalent.

**Comment by adidahiya** [comment]
that's a good point @etsuo... the rule should probably make an exception for when the first argument to Object.assign is a `new SomeConstructor(...)` expression because spread is not functionally equivalent

**Comment by danvk** [comment]
Is the issue just with new? Or with any object with a prototype other than
Object.prototype?
